src/category_theory/limits/kan_extension.lean

```diff
@@ -0,0 +1,370 @@
+/-
+Copyright (c) 2021 Adam Topaz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta, Adam Topaz
+-/
+import category_theory.punit
+import category_theory.comma
+import category_theory.limits.functor_category
+import category_theory.limits.shapes.terminal
+
+/-!
+
+# Kan extensions
+
+This file defines the right and left Kan extensions of a functor.
+They exist under the assumption that the target category has enough limits
+resp. colimits.
+
+The main definitions are `Ran ι` and `Lan ι`, where `ι : S ⥤ L` is a functor.
+Namely, `Ran ι` is the right Kan extension, while `Lan ι` is the left Kan extension,
+both as functors `(S ⥤ D) ⥤ (L ⥤ D)`.
+
+To access the right resp. left adjunction associated to these, use `Ran.adjunction`
+resp. `Lan.adjunction`.
+
+# Projects
+
+A lot of boilerplate could be generalized by defining and working with pseudofunctors.
+
+-/
+
+noncomputable theory
+
+namespace category_theory
+
+open limits
+
+universes v u₁ u₂ u₃
+
+variables {S : Type v} {L : Type u₂} {D : Type u₃}
+variables [category.{v} S] [category.{v} L] [category.{v} D]
+variables (ι : S ⥤ L)
+
+namespace Ran
+
+local attribute [simp] comma.snd comma.map_left
+
+/-- The index category of limits used to define `Ran`. -/
+@[simp, derive category, nolint has_inhabited_instance]
+def index (l : L) := comma (functor.from_punit l) ι
```

Comment:
I wonder if it is time to factor this out. We use the pattern `comma (functor.from_punit X) F` in four existing places in mathlib. Maybe there is not that much to say about it, except the next 6 (or 7?) statements.

The only name I can [find](https://en.wikipedia.org/wiki/Comma_category#Other_variations) for this is `structured_arrow`.

Reply:
I guess a counter argument is that it is time for pseudofunctors...

Reply:
I definitely agree it should be factored out - maybe even between `comma` and `over`.

Reply:
One place it's important to talk about this category is for the adjoint functor theorems - for instance the existence of a left adjoint to a functor G : D -> C is equivalent to an initial object of (A, G) for every object A of C; and we have that the forgetful functor (A, G) -> D creates limits of any shape which G preserves; so the adjoint functor theorems are essentially just asking for conditions on (A, G) on top of having limits to give it an initial object as well. So essentially properties of the category `comma (functor.from_punit A) G` are key steps in the proof (eg #4885)

Reply:
Okay, I've factored this out as #6830. If no one objects, I can add `costructured_arrow` there, and also refactor the existing unnamed used in mathlib. Hopefully it is easy to make this PR build on that?